src/gallium/drivers/lima/lima_draw.c

```diff
@@ -376,14 +378,33 @@ lima_pack_plbu_cmd(struct lima_context *ctx, const struct pipe_draw_info *info)
    plbu_cmd[i++] = fui(ctx->viewport.far);
    plbu_cmd[i++] = 0x1000010F; /* DEPTH_RANGE_FAR */
 
+   struct pipe_resource *indexbuf = NULL;
+   struct pipe_draw_info new_info;
    if (info->index_size) {
       plbu_cmd[i++] = gl_position_va;
       plbu_cmd[i++] = 0x10000100; /* INDEXED_DEST */
 
       struct lima_resource *res = lima_resource(info->index.resource);
+      if (info->has_user_indices) {
+         unsigned index_offset = 0;
+         if (!util_upload_index_buffer(&ctx->base, info, &indexbuf,
+                                       &index_offset)) {
+            fprintf(stderr, "index buffer upload error\n");
+            return;
+         }
+
+         new_info = *info;
```

Comment:
I think you don't need a new draw_info and res->index_offset here, just a lima_resource and initialized to 0 offset is enough. But do remember to call lima_submit_add_bo before free this indexbuf, otherwise this buffer may be freed before lima_flush (BTW. I forgot to call it either here).

Reply:
The new draw_info is required because this arrives here as a pointer to const data, so we can't pass that to util_upload_index_buffer. I took this pattern from the freedreno driver.
The index_offset is required here because the resource returned in indexbuf is a buffer with some other stuff in it, apparently vertex data, and the index buffer is not right at res->bo->va. It took quite some time to figure out why it was not working in the beginning, but by dumping the buffers I saw that the indices are actually in an offset and index_offset contains the right offset to reach it.
I'm saving it in the resource for the case where the buffer was already uploaded and info->has_user_indices is not set. Not sure if saving it in the bo or resource is better though.

I'll check the missing lima_submit_add_bo, maybe that is what causes the bug I mentioned.

Reply:
How do you save the resource for next reuse? Seems this resource is freed after this function. Also this new_info won't change content of *info and just freed after this function.

I think it should be like this:
```
struct pipe_resource *indexbuf = NULL;
unsigned index_offset = 0;
struct lima_resource *res;
if (info->has_user_indices) {
   util_upload_index_buffer(&ctx->base, info, &indexbuf, &index_offset);
   res = lima_resource(indexbuf);
}
else
   res = lima_resource(info->index.resource);
lima_bo_update(res->bo, false, true);
lima_submit_add_bo(ctx->gp_submit, res->bo, LIMA_SUBMIT_READ);
plbu_cmd[i++] = res->bo->va + info->start + info->start_instance * info->index_size + index_offset;
if (indexbuf)
   pipe_resource_reference(&indexbuf, NULL);
```